copier/jinja_ext.py

```diff
@@ -0,0 +1,109 @@
+"""Jinja2 extensions built for Copier."""
+
+from __future__ import annotations
+
+from typing import Any, Callable, Sequence
+
+from jinja2 import nodes
+from jinja2.exceptions import UndefinedError
+from jinja2.ext import Extension
+from jinja2.parser import Parser
+from jinja2.sandbox import SandboxedEnvironment
+
+
+class YieldEnvironment(SandboxedEnvironment):
+    """Jinja2 environment with a `yield_context` attribute.
+
+    This is simple environment class that extends the SandboxedEnvironment
+    for use with the YieldExtension, mainly for avoiding type errors.
+
+    We use the SandboxedEnvironment because we want to minimize the risk of hidden malware
+    in the templates so we use the SandboxedEnvironment instead of the regular one.
+    Of course we still have the post-copy tasks to worry about, but at least
+    they are more visible to the final user.
+    """
+
+    yield_context: dict[str, Any]
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.extend(yield_context=dict())
+
+
+class YieldExtension(Extension):
+    """`Jinja2 extension for the `yield` tag.
+
+    If `yield` tag is used in a template, this extension sets the `yield_context` attribute to the
+    jinja environment. `yield_context` is a dictionary with the following keys:
+    - `single_var`: The name of the variable that will be yielded.
+    - `looped_var`: The variable that will be looped over.
+
+    Note that this extension just sets the `yield_context` attribute but renders template
+    as usual. It is caller's responsibility to use the `yield_context` attribute in the
+    template to generate the desired output.
+
+    Example:
+        template: "{% yield single_var from looped_var %}"
+        context: {"looped_var": [1, 2, 3], "single_var": "item"}
+
+        then,
+        >>> from copier.jinja_ext import YieldEnvironment, YieldExtension
+        >>> env = YieldEnvironment(extensions=[YieldExtension])
+        >>> template = env.from_string("{% yield single_var from looped_var %}{{ single_var }}{% endyield %}")
+        >>> template.render({"looped_var": [1, 2, 3]})
+        ''
+        >>> env.yield_context
```

Comment:
Following the suggestion above:

```suggestion
        >>> env.yield_name, env.yield_iterable
```